File: hosts/android/feature/products/impl/src/main/java/io/paritytech/polkadotapp/feature_products_impl/domain/pocketAddCard/PocketAddCardInteractor.kt
package io.paritytech.polkadotapp.feature_products_impl.domain.pocketAddCard

import android.net.Uri
import io.paritytech.polkadotapp.common.utils.flatMap
import io.paritytech.polkadotapp.feature_products_api.domain.pocket.PocketCard
import io.paritytech.polkadotapp.feature_products_api.domain.pocket.PocketCardId
import io.paritytech.polkadotapp.feature_products_api.domain.pocket.PocketCardKey
import io.paritytech.polkadotapp.feature_products_api.model.JsImageSource
import io.paritytech.polkadotapp.feature_products_api.model.JsWidget
import io.paritytech.polkadotapp.feature_products_api.model.ProductId
import io.paritytech.polkadotapp.feature_products_impl.domain.pocket.CachedPocketCard
import io.paritytech.polkadotapp.feature_products_impl.domain.pocket.PocketCardStore
import io.paritytech.polkadotapp.feature_products_impl.domain.pocket.PocketImageResolver
import io.paritytech.polkadotapp.feature_products_impl.domain.pocket.PocketPreviewLoader
import io.paritytech.polkadotapp.feature_products_impl.domain.pocket.PublishedPocketCards
import javax.inject.Inject

/** A published card as the approval sheet shows it: the face the user approves is the one that is stored. */
data class PocketAddCardOffer(
    val key: PocketCardKey,
    val productName: String,
    val title: String,
    val face: JsWidget,
)

class PocketAddCardInteractor @Inject constructor(
    private val publishedCards: PublishedPocketCards,
    private val previewLoader: PocketPreviewLoader,
    private val store: PocketCardStore,
    private val images: PocketImageResolver,
) {
    suspend fun loadOffer(productId: ProductId, cardId: PocketCardId): Result<PocketAddCardOffer> =
        publishedCards.find(productId, cardId).flatMap { published ->
            previewLoader.load(productId, published.definition).map { face ->
                PocketAddCardOffer(
                    key = PocketCardKey(productId, cardId),
                    productName = published.product.name,
                    title = published.definition.title,
                    face = face,
                )
            }
        }

    suspend fun resolveFaceImage(productId: ProductId, source: JsImageSource): Result<Uri> =
        images.resolve(productId, source)

    suspend fun approve(offer: PocketAddCardOffer): Result<Unit> = runCatching {
        store.addCard(CachedPocketCard(PocketCard(offer.key, offer.title, privileged = false), offer.face))
    }
}

File: hosts/ios/polkadot-app/Modules/Products/Pocket/PocketAddCardInteractor.swift
import Foundation
import PolkadotUI
import Products
import TrUAPIHost

enum PocketPublishError: Error, Equatable, CustomStringConvertible {
    case noPocket
    case unknownCard

    var description: String {
        switch self {
        case .noPocket: "this product publishes no Pocket cards"
        case .unknownCard: "this product publishes no such card"
        }
    }
}

/// A card as its product's worker manifest publishes it. The face is read from
/// the worker's own archive, which resolves under `workerContentId` rather than
/// under the product's base name.
struct PublishedPocketCard {
    let productId: ProductId
    let productName: String
    let workerContentId: ProductId
    let definition: PocketCardDefinition
}

/// Looks a card up in the worker manifest of the product that claims to back it.
protocol PublishedPocketCardsResolving: Sendable {
    func find(productId: ProductId, cardId: PocketCardId) async throws -> PublishedPocketCard
}

/// A published card as the approval sheet shows it.
struct PocketAddCardOffer {
    let key: PocketCardKey
    let productName: String
    let title: String
    let face: RendererNode
}

/// Loads what the user is being asked to approve, and stores exactly that.
///
/// The offer is loaded once: the face shown in the sheet is the face kept for
/// the card, so what the user approved is what the Pocket draws.
struct PocketAddCardInteractor {
    private let publishedCards: any PublishedPocketCardsResolving
    private let previews: PocketPreviewLoader
    private let store: any PocketCardStore

    init(
        publishedCards: any PublishedPocketCardsResolving,
        previews: PocketPreviewLoader,
        store: any PocketCardStore
    ) {
        self.publishedCards = publishedCards
        self.previews = previews
        self.store = store
    }

    func loadOffer(productId: ProductId, cardId: PocketCardId) async throws -> PocketAddCardOffer {
        let published = try await publishedCards.find(productId: productId, cardId: cardId)
        let face = try await previews.load(
            contentId: published.workerContentId,
            preview: published.definition.preview
        )

        return PocketAddCardOffer(
            key: PocketCardKey(productId: published.productId, cardId: cardId),
            productName: published.productName,
            title: published.definition.title,
            face: face
        )
    }

    /// The card is stored with the face the sheet showed, so what the Pocket
    /// draws next is what the user approved. Only the host places a privileged
    /// card, so an approved one never is.
    func approve(_ offer: PocketAddCardOffer) async {
        await store.add(
            PocketCardEntry(key: offer.key, title: offer.title, privileged: false),
            face: offer.face
        )
    }
}
